### PR TITLE
Add common validation for compression and encryption

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add common validation for compression and encryption

## Motivation and Context
@Praveenrajmani suggested this change while reviewing #7963 - it makes sense to split the patch as a separate PR. 

## How to test this PR?
Just set the compression parameters while running gateway should fail. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#1f554fd2-a7ea-46aa-ba06-a649fbb4ef2e) |
| mint-dist-xl.sh | :x: [more...](#a87fddf9-c316-4e25-b43f-e49ad0c905f0) |
<a name="a87fddf9-c316-4e25-b43f-e49ad0c905f0"></a>
### 7978-77a341b/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31633
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b8d42c4f0f93:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 1 minutes and 12 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 39 seconds
(8/14) Running minio-dotnet tests ... done in 43 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 21 seconds
(10/14) Running minio-java tests ... FAILED in 1 minutes and 11 seconds
{
  "name": "minio-java",
  "function": "listIncompleteUploads(final String bucketName, final String prefix, final boolean recursive)",
  "args": "prefix: minio, recursive: true",
  "duration": 420,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=SlowDown, message=Please reduce your request, bucketName=minio-java-test-h0l8hd, objectName=minio-java-test-1e4ed5, resource=/minio-java-test-h0l8hd/minio-java-test-1e4ed5, requestId=15B4D19647858091, hostId=ba01614c-01cb-479e-8ea7-393e978579f8)\nrequest={method=DELETE, url=http://72.28.97.61:31633/minio-java-test-h0l8hd/minio-java-test-1e4ed5?uploadId=bad46f4a-32de-4968-9ceb-948539ccada3, headers=Host: 72.28.97.61:31633\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20190726T014333Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190726/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 359\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.2019-07-26T01-24-33Z\nVary: Origin\nX-Amz-Request-Id: 15B4D19647858091\nX-Xss-Protection: 1; mode=block\nDate: Fri, 26 Jul 2019 01:43:34 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1204), io.minio.MinioClient.execute(MinioClient.java:1070), io.minio.MinioClient.executeDelete(MinioClient.java:1347), io.minio.MinioClient.abortMultipartUpload(MinioClient.java:5149), io.minio.MinioClient.putObject(MinioClient.java:4337), io.minio.MinioClient.putObject(MinioClient.java:4183), FunctionalTest.listIncompleteUploads_test3(FunctionalTest.java:1679), FunctionalTest.runTests(FunctionalTest.java:2990), FunctionalTest.main(FunctionalTest.java:3118)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="1f554fd2-a7ea-46aa-ba06-a649fbb4ef2e"></a>
### 7978-77a341b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:32688
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 665bafe8e5f5:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 13 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ujkwrnfb66f591",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 12527,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:32688/aws-sdk-go-test-ujkwrnfb66f591: dial tcp 72.28.97.57:32688: connect: no route to host",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

File: cmd/common-main.go [hunk 1]
```diff
@@ -38,6 +38,29 @@ import (
 	xnet "github.com/minio/minio/pkg/net"
 )
 
+func verifyObjectLayerFeatures(name string, objAPI ObjectLayer) {
+	// Encryption support checks.
+	{
```
Comment (aead) [inline cmd/common-main.go]: We can remove the extra brackets.

**Review by aead — APPROVED** [review]

**Review by Praveenrajmani — APPROVED** [review]
LGTM

File: cmd/common-main.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/common-main.go]: Done removed..

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#754498a9-c6e4-41c1-9653-574d8b779fc9) |
| mint-large-bucket.sh | :x: [more...](#c09a113e-f901-4c67-8fdd-23a0cf7302b4) |
| mint-fs.sh | :x: [more...](#105581e4-4518-4a69-9495-a9429de12e3e) |
| mint-dist-xl.sh | :x: [more...](#42b62575-beb5-438c-a7c8-64139d136fcd) |
<a name="c09a113e-f901-4c67-8fdd-23a0cf7302b4"></a>
### 7978-078f7b7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31782
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c8327c420c2a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 48 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 2 minutes and 16 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 2 minutes and 5 seconds
{
  "name": "mc",
  "duration": "98225",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-25499 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-7013  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-7013/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-7013/datafile-0-b`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-7013/datafile-1-b`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-7013/datafile-1.03-MB`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-7013/datafile-1-kB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-7013/datafile-100-kB`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-7013/datafile-10-MB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-7013/datafile-129-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-7013/datafile-11-MB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-7013/datafile-10-kB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-7013/datafile-5-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-7013/datafile-5243880-b`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-7013/datafile-6-MB`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-7013/datafile-65-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-7013/datafile-33-kB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-11-MB`. Please reduce your request\nTotal: 233.17 MiB, Transferred: 423.68 MiB, Speed: 4.32 MiB/s"
}

Executed 6 out of 14 tests successfully.
```
<a name="42b62575-beb5-438c-a7c8-64139d136fcd"></a>
### 7978-078f7b7/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:31657
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1bc6ae45adc0:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 1 minutes and 11 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 16 seconds
{
  "name": "mc",
  "duration": "7299",
  "function": "test_mirror_list_objects",
  "status": "FAIL",
  "error": "mirror and list differs"
}

Executed 6 out of 14 tests successfully.
```
<a name="105581e4-4518-4a69-9495-a9429de12e3e"></a>
### 7978-078f7b7/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30776
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a4f7850a024b:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 1 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 4 seconds
{
  "name": "mc",
  "duration": "1109",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-22113 --quiet --no-color cp myminio/mc-test-bucket-20381/mc-test-object-28656 mc-test-object-28656.downloaded  >>> \n`myminio/mc-test-bucket-20381/mc-test-object-28656` -> `mc-test-object-28656.downloaded`\nmc: <ERROR> Failed to copy `http://72.28.97.55:30776/mc-test-bucket-20381/mc-test-object-28656`. read tcp 172.17.0.2:59402->72.28.97.55:30776: read: connection reset by peer\nmc: <ERROR> Session safely terminated. To resume session `mc session resume jHUlymXA`"
}

Executed 6 out of 14 tests successfully.
```
<a name="754498a9-c6e4-41c1-9653-574d8b779fc9"></a>
### 7978-078f7b7/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31751
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e3bd02143890:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 13 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-dko5rdsi3k1u66",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 12566,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:31751/aws-sdk-go-test-dko5rdsi3k1u66: dial tcp 72.28.97.57:31751: connect: no route to host",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#7f25f8f8-70de-478a-932d-d48c8bb8ec07) |
| mint-dist-xl.sh | :x: [more...](#16c85209-afce-46f1-acdc-91956bbd0d14) |
<a name="16c85209-afce-46f1-acdc-91956bbd0d14"></a>
### 7978-a7f60d3/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30916
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e8cf9768f896:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 2 minutes and 5 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 56 seconds
(8/14) Running minio-dotnet tests ... done in 42 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 23 seconds
(10/14) Running minio-java tests ... FAILED in 1 minutes and 12 seconds
{
  "name": "minio-java",
  "function": "presignedGetObject(String bucketName, String objectName)",
  "duration": 71,
  "status": "FAIL",
  "error": "java.lang.Exception: failed to read object. Response: Response{protocol=http/1.1, code=503, message=Service Unavailable, url=http://72.28.97.52:30916/minio-java-test-15as1ql/minio-java-test-35or1qq?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190726%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190726T102610Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=7b4e2c0f1b902547b88c301d3630fab7e4471890af070484a067d642219124b5}, Response body: <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>minio-java-test-35or1qq</Key><BucketName>minio-java-test-15as1ql</BucketName><Resource>/minio-java-test-15as1ql/minio-java-test-35or1qq</Resource><RequestId>15B4EE1AEDF455B3</RequestId><HostId>91fcfd52-9bfa-4769-bcf4-efe8c1da9969</HostId></Error> >>> [FunctionalTest.readObject(FunctionalTest.java:201), FunctionalTest.presignedGetObject_test1(FunctionalTest.java:1767), FunctionalTest.runTests(FunctionalTest.java:2994), FunctionalTest.main(FunctionalTest.java:3118)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="7f25f8f8-70de-478a-932d-d48c8bb8ec07"></a>
### 7978-a7f60d3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:30414
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 644b63f19449:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 52 seconds
(4/14) Running aws-sdk-ruby tests ... done in 15 seconds
(5/14) Running awscli tests ... done in 2 minutes and 18 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 2 minutes and 19 seconds
{
  "name": "mc",
  "duration": "94715",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-16901 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-17120  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-17120/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-17120/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-17120/datafile-1-kB`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-17120/datafile-10-MB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-17120/datafile-1.03-MB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-17120/datafile-1-b`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-17120/datafile-10-kB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-17120/datafile-100-kB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-17120/datafile-11-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-17120/datafile-33-kB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-17120/datafile-129-MB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-17120/datafile-5-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-17120/datafile-5243880-b`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-17120/datafile-65-MB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-17120/datafile-6-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-10-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Please reduce your request\nTotal: 233.17 MiB, Transferred: 482.61 MiB, Speed: 5.10 MiB/s"
}

Executed 6 out of 14 tests successfully.
```

**Pull request merged**